### DISCUSSION TITLE
Trouble getting semantic highlighting + Color Identifiers working together with simple language grammar

### DISCUSSION THREAD

**Discussion by bigyihsuan** [microsoft/vscode-discussions#1034] [Extension Development QnA] [answered]
I'm trying to get the [Color Identifiers extension](https://marketplace.visualstudio.com/items?itemName=MatthewNespor.vscode-color-identifiers-mode) working with a small syntax highlighting extension I'm making (https://github.com/bigyihsuan/pmd-mml-syntax-highlighting).

The language is the PMD flavor of MML for the NEC PC-98. What I want to do is to have each note (in general something maxing regexp `([abcdefgr])([\$\%]?)(\d*))` be considered as a variable for Color Identifiers to color.

However, Color Identifiers seems to be not picking up on it. Below, every single character that is red should not be red:

![image](https://github.com/microsoft/vscode-discussions/assets/12767476/56c9a05f-b3d4-4fa3-a8a4-cedba4c8a078)

The code for this pattern is here: https://github.com/bigyihsuan/pmd-mml-syntax-highlighting/blob/79e52d40a943d7e12093e36dd831f77bba71aa61/syntaxes/pmdmml.tmLanguage.json#L123-L141

I've tried following the example on the [Semantic Highlighting Guide](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide), but all I get is how the type `SemanticTokensLegend` and `DocumentSemanticTokensProvider` and such don't exist.

How can I get the semantic highlighting working?

**Comment by tamuratak** [comment, marked as answer]
What your extension offers is Syntax Highlighting with [TextMate grammar](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#textmate-grammars), not Semantic Highlighting. They are different ones. If you want to provide Semantic Highlighting, see a [sample](https://github.com/microsoft/vscode-extension-samples/tree/main/semantic-tokens-sample). You have to write [codes](https://github.com/microsoft/vscode-extension-samples/blob/5d892ccedd9f162094bd7f519fda1fd07e28f2bb/semantic-tokens-sample/src/extension.ts#L67-L94) to parse files with TypeScript by yourself.

**Reply by bigyihsuan** [reply]
Oh, so I'll need manually implement a lexer+parser for semantic highlighting to work?

**Reply by tamuratak** [reply]
Yes.

If a tree-sitter parser exists for the language you are interested in, you could reuse the tree-sitter parser like [vscode-anycode](https://github.com/microsoft/vscode-anycode).

**Reply by bigyihsuan** [reply]
Hm. MML doesn't have much syntax in common with other languages. I might have to hand-roll this one.

**Comment by bigyihsuan** [comment]
Well I found a fix to one of my problems.

The reason why the types weren't found was because I was using the `vscode` package instead of the `@types/vscode` package.
